apps/api/src/drizzle/transaction.interceptor.ts

```diff
@@ -0,0 +1,52 @@
+import {
+    CallHandler,
+    ExecutionContext,
+    Inject,
+    Injectable,
+    NestInterceptor,
+} from "@nestjs/common";
+import { GqlExecutionContext } from "@nestjs/graphql";
+import { drizzle } from "drizzle-orm/node-postgres";
+import { Pool } from "pg";
+import { Observable, from, throwError } from "rxjs";
+import { catchError, switchMap } from "rxjs/operators";
+import { CONNECTION_POOL } from "./drizzle.module-definition";
+import * as schema from "./schema";
+
+export const DRIZZLE_DATABASE_KEY = "DRIZZLE_DATABASE";
+
+@Injectable()
+export class TransactionInterceptor implements NestInterceptor {
+    constructor(@Inject(CONNECTION_POOL) private readonly pool: Pool) {}
```

Comment:
Errm... iirc I think we don't inject db this way :thinking:

Reply:
hmm
I think I'm missing something here @xUser5000

Reply:
Yes, I'm aware that this is not the normal way to inject the drizzle service.
The problem if I do it in the normal way (`private readonly drizzleService: DrizzleService`) is that the transaction interceptor now depends on the drizzle service, which in turn exposes something (drizzle db/transaction instance) that should be set by the transaction interceptor itself... circular dependency

In order to resolve this cycle, I had to inject the same primitive (the connection pool) used by the drizzle service to create the db/transaction instance.

Reply:
Although, you pointed out something that I overlooked. I don't really know if creating multiple `drizzle()` instances can cause some resource leaks.
I'll investigate that

Reply:
@xUser5000 
btw, will using `AsyncLocalStorage` help in regards to eliminating the need for `request scoped` _everything_?

Reply:
@G0maa 
I asked claud 3.5 sonnet about the code and whether or not it has resource leaks. It gave me this:

---
To evaluate potential resource leaks in this code, let's analyze the key components:

Connection Pool:
The Pool from the pg library is created in the DrizzleModule and injected into the DrizzleService. Pools are designed to manage and reuse connections efficiently, which helps prevent leaks.
DrizzleService Scope:
The DrizzleService is scoped to REQUEST, meaning a new instance is created for each request. This doesn't directly cause leaks, but it's important to ensure resources are properly managed.
Drizzle ORM Usage:
The drizzle function is called in the db getter of DrizzleService. It either returns an existing database instance from the context or creates a new one.

Potential Concerns:

Multiple Drizzle Instances:
If the db getter is called multiple times within a request and DRIZZLE_DATABASE_KEY is not set in the context, it could create multiple drizzle instances. However, this doesn't necessarily lead to leaks as long as the underlying Pool is managed correctly.
Transaction Management:
The presence of TransactionInterceptor suggests that transactions are being managed, which is good for resource handling. However, we'd need to see its implementation to ensure it's correctly releasing resources.
Pool Shutdown:
There's no visible code for shutting down the Pool when the application terminates. This could potentially lead to connection leaks if not handled elsewhere.

Conclusion:
Based on the provided code, there are no immediate, obvious resource leaks. However, to ensure complete resource safety:

Ensure the Pool is properly closed when the application shuts down.
Verify that the TransactionInterceptor correctly manages database transactions and releases resources.
Consider caching the drizzle instance in the DrizzleService to avoid creating multiple instances per request.

To be absolutely certain, you'd need to test the application under load and monitor for any connection or memory leaks. Also, reviewing the implementation of TransactionInterceptor and any application shutdown hooks would provide a more complete picture.
---

So, in conclusion, I think we're safe when it comes to database connection leaks or anything like that.

Reply:
Regaring `AsyncLocalStorage`, you are right, we can potentially throw away request-scoped providers and replace and make them access the request object through `AsyncLocalStorage`.

I'll investigate that further and update the PR for another review

Reply:
Investigation results: spent 3 hours trying to make `AsyncLocalStorage` work alongside the `GraphQLModule` and wasn't lucky enough ig

However, Claude suggested another solution to mitigate the problem of all providers being request-scoped. In a nutshell, we can create a "proxy" provider around the drizzle service that is not request-scoped and does a little bit of magic to retrieve a correct instance of the drizzle service. I'll try to look into it.